File: examples/python/input/textarea_editor.py
#!/usr/bin/env python3
"""Multi-line entry and the external-editor hook.

    make run-example EX=python/input/textarea_editor

Keys: Enter inserts a newline, Ctrl-D submits, Ctrl-G opens $EDITOR.
"""

from __future__ import annotations

import sparcli as sc


def run_textarea() -> None:
    """Boxed multi-line editor with soft-wrapped long lines."""
    notes = sc.textarea('Release notes (Ctrl-D submits)', sc.TextareaOpts(
        placeholder='What changed?', boxed=True,
        border=sc.BorderStyle(sc.BorderType.ROUNDED), width=52))
    if notes is not None:
        sc.println(f'  -> {len(notes.splitlines())} line(s), {len(notes)} bytes')


def run_editor_input() -> None:
    """External editor: Ctrl-G opens $EDITOR; text_input collapses newlines."""
    message = sc.text_input('Commit message (Ctrl-G opens the editor)',
                            sc.TextInputOpts(external_editor=True))
    if message is not None:
        sc.println(f'  -> {message!r}')


def main() -> None:
    if not sc.input_available():
        sc.alert.warning('Run this example in a real terminal (not piped).')
        return
    run_textarea()
    run_editor_input()


if __name__ == '__main__':
    main()

File: examples/rust/input/textarea_editor.rs
//! Multi-line entry and the external-editor hook.
//!
//!   cargo run -p sparcli --example input_textarea_editor
//!
//! Keys: Enter inserts a newline, Ctrl-D submits, Ctrl-G opens $EDITOR.

use sparcli::*;

fn main() -> sparcli::Result<()> {
    if !input_available() {
        alert::warning("Run this example in a real terminal (not piped).");
        return Ok(());
    }

    // Boxed multi-line editor with soft-wrapped long lines.
    if let Some(notes) =
        textarea("Release notes (Ctrl-D submits)", TextareaOpts::new().boxed(52))?
    {
        println(&format!("  -> {} bytes", notes.len()), Style::default());
    }

    // External editor: Ctrl-G opens $EDITOR; text_input collapses newlines.
    if let Some(message) = text_input(
        "Commit message (Ctrl-G opens the editor)",
        TextInputOpts::new().external_editor(true),
    )? {
        println(&format!("  -> {message:?}"), Style::default());
    }

    Ok(())
}
